<a href="https://colab.research.google.com/github/akifkashif/Black-Coffer-Blogs-Sentiment-Analysis/blob/main/Akif_Black_Coffer_Project_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#importing library
import pandas as pd
import requests
from bs4 import BeautifulSoup
import numpy as np
from google.colab import files

In [ ]:
#importing google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#importing input file
df_project=pd.read_csv('/content/drive/MyDrive/Input.csv')[['URL_ID','URL']]

In [ ]:
#extracting the first 5 rows of the dataset
df_project.head()

URL_ID                                                URL
0     1.0  https://insights.blackcoffer.com/is-telehealth...
1     2.0  https://insights.blackcoffer.com/how-telehealt...
2     3.0  https://insights.blackcoffer.com/is-telemedici...
3     4.0  https://insights.blackcoffer.com/is-telehealth...
4     5.0  https://insights.blackcoffer.com/how-people-di...

In [ ]:
#extracting the last 5 rows of the dataset
df_project.tail()

URL_ID  URL
1014     NaN  NaN
1015     NaN  NaN
1016     NaN  NaN
1017     NaN  NaN
1018     NaN  NaN

#Here there are null values starting from 151 row of the dataset so capping the rows till 150 should be done 

In [ ]:
#getting the number of rows and columns
df_project.shape

(1019, 2)

In [ ]:
#removing the row which had null values in them
df_project=df_project.iloc[0:150]

In [ ]:
#viewing data in dataframe
df_project['URL'][0]

'https://insights.blackcoffer.com/is-telehealth-the-future-of-healthcare-3/'

# Data extraction

In [ ]:
#performing data extraction on a single URL
j=df_project['URL'][40]
   
headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36'}#giving user access
page=requests.get(j,headers=headers)#loading text in url
soup=BeautifulSoup(page.content,'html.parser')#parsing url text
content=soup.findAll(attrs={'class':'td-post-content'})#extracting only text part
content=content[0].text.replace('<p>'," ").replace('\n'," ").replace('\xa0',' ')#replace end line symbol with space 
title=soup.findAll(attrs={'class':'entry-title'})#extracting title of website
print(title)

[<h3 class="entry-title td-module-title"><a href="https://insights.blackcoffer.com/ranking-customer-behaviours-for-business-strategy/" rel="bookmark" title="Ranking customer behaviours for business strategy">Ranking customer behaviours for business strategy</a></h3>, <h3 class="entry-title td-module-title"><a href="https://insights.blackcoffer.com/algorithmic-trading-for-multiple-commodities-markets-like-forex-metals-energy-etc/" rel="bookmark" title="Algorithmic trading for multiple commodities markets, like Forex, Metals, Energy, etc.">Algorithmic trading for multiple commodities markets, like Forex, Metals, Energy, etc.</a></h3>, <h3 class="entry-title td-module-title"><a href="https://insights.blackcoffer.com/trading-bot-for-forex/" rel="bookmark" title="Trading Bot for FOREX">Trading Bot for FOREX</a></h3>, <h3 class="entry-title td-module-title"><a href="https://insights.blackcoffer.com/python-model-for-the-analysis-of-sector-specific-stock-etfs-for-investment-purposes%ef%bf%bc/"

In [ ]:
#extracting text from all the url
main_df = pd.DataFrame()
for i in range(len(df_project)):
  #  print(i) 
   j=df_project['URL'][i]
   
   headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36'}#giving user access
   page=requests.get(j,headers=headers)#loading text in url
   soup=BeautifulSoup(page.content,'html.parser')#parsing url text
   content=soup.findAll(attrs={'class':'td-post-content'})#extracting only text part
   try:  
    content=content[0].text.replace('<p>'," ").replace('\n'," ").replace('\xa0',' ')#replace end line symbol with space 
    title=soup.findAll(attrs={'class':'entry-title'})#extracting title of website

    title=title[0].text.replace('\n',"  ").replace('/',"")
    text=title+ '.' +content #merging title and content text
    df1=pd.DataFrame(pd.Series(text),columns=['data'])#creating series data frame
    df1['URL'] = j
    main_df = pd.concat([df1,main_df],ignore_index=True)
   
   except:
     print(i)
     continue


   

23
43
56
143


#Here the row numbers 23, 43, 56, 143 have "page not found error" in their urls so in total after extraction we have 146 url for analysis

In [ ]:
main_df.head()

data  \
0  Ranking customer behaviours for business strat...   
1  Ranking customer behaviours for business strat...   
2  Ranking customer behaviours for business strat...   
3  Ranking customer behaviours for business strat...   
4  Ranking customer behaviours for business strat...   

                                                 URL  
0  https://insights.blackcoffer.com/challenges-an...  
1  https://insights.blackcoffer.com/business-anal...  
2  https://insights.blackcoffer.com/big-data-anal...  
3  https://insights.blackcoffer.com/the-future-of...  
4  https://insights.blackcoffer.com/blockchain-fo...

In [ ]:
main_df = main_df.loc[:,['URL','data']]

In [ ]:
main_df.head()

URL  \
0  https://insights.blackcoffer.com/challenges-an...   
1  https://insights.blackcoffer.com/business-anal...   
2  https://insights.blackcoffer.com/big-data-anal...   
3  https://insights.blackcoffer.com/the-future-of...   
4  https://insights.blackcoffer.com/blockchain-fo...   

                                                data  
0  Ranking customer behaviours for business strat...  
1  Ranking customer behaviours for business strat...  
2  Ranking customer behaviours for business strat...  
3  Ranking customer behaviours for business strat...  
4  Ranking customer behaviours for business strat...

In [ ]:
#visualizing shape of the dataframe 
main_df.shape

(146, 2)

# Data analysis

In [ ]:
#information of data frame
main_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146 entries, 0 to 145
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   URL     146 non-null    object
 1   data    146 non-null    object
dtypes: object(2)
memory usage: 2.4+ KB


In [ ]:
main_df['data']

0      Ranking customer behaviours for business strat...
1      Ranking customer behaviours for business strat...
2      Ranking customer behaviours for business strat...
3      Ranking customer behaviours for business strat...
4      Ranking customer behaviours for business strat...
                             ...                        
141    Ranking customer behaviours for business strat...
142    Ranking customer behaviours for business strat...
143    Ranking customer behaviours for business strat...
144    Ranking customer behaviours for business strat...
145    Ranking customer behaviours for business strat...
Name: data, Length: 146, dtype: object

In [ ]:
#creating new dataframe
df = main_df.copy()

In [ ]:
#converting text to sentence
import re
#splitting text on '.'
df['data']= main_df['data'].apply(lambda x: x.split('([\.]\s)'))


In [ ]:
df.head()

URL  \
0  https://insights.blackcoffer.com/challenges-an...   
1  https://insights.blackcoffer.com/business-anal...   
2  https://insights.blackcoffer.com/big-data-anal...   
3  https://insights.blackcoffer.com/the-future-of...   
4  https://insights.blackcoffer.com/blockchain-fo...   

                                                data  
0  [Ranking customer behaviours for business stra...  
1  [Ranking customer behaviours for business stra...  
2  [Ranking customer behaviours for business stra...  
3  [Ranking customer behaviours for business stra...  
4  [Ranking customer behaviours for business stra...

In [ ]:
#removing . char from each rows
def abcd(x):    
    nopunc =[char for char in x if char != '.']
    return ''.join(nopunc)
df['data']=df['data'].apply(abcd)


In [ ]:
#replacing empty space with null values
df=df.replace('',np.nan,regex=True)
df=df.mask(df==" ")
df=df.dropna()
df.reset_index(drop=True,inplace=True)

In [ ]:
#importing nltk library and stopwords
import nltk
import string

In [ ]:
punc=[punc for punc in string.punctuation]

In [ ]:
#importing stop words files that are provided
SW_Auditor=pd.read_csv("/content/drive/MyDrive/20211030 Test Assignment/StopWords/StopWords_Auditor.txt",header=None)
SW_Currencies=pd.read_csv("/content/drive/MyDrive/20211030 Test Assignment/StopWords/StopWords_Currencies.txt",header=None,encoding="ISO-8859-1",sep='\n' )
SW_DatesandNumbers=pd.read_csv("/content/drive/MyDrive/20211030 Test Assignment/StopWords/StopWords_DatesandNumbers.txt",header=None)
SW_Generic=pd.read_csv("/content/drive/MyDrive/20211030 Test Assignment/StopWords/StopWords_Generic.txt",header=None)
SW_GenericLong=pd.read_csv("/content/drive/MyDrive/20211030 Test Assignment/StopWords/StopWords_GenericLong.txt",header=None)
SW_Geographic=pd.read_csv("/content/drive/MyDrive/20211030 Test Assignment/StopWords/StopWords_Geographic.txt",header=None)
SW_Names=pd.read_csv("/content/drive/MyDrive/20211030 Test Assignment/StopWords/StopWords_Names.txt",header=None)


In [ ]:
#creating func for removing stop words
def text_process(text):
    nopunc =[char for char in text if char not in punc or char not in [':',',','(',')','’','?']]
    nopunc=''.join(nopunc)
    txt=' '.join([word for word in nopunc.split() if word.lower() not in SW_Auditor])
    txt1=' '.join([word for word in txt.split() if word.lower() not in SW_Currencies])
    txt2=' '.join([word for word in txt1.split() if word.lower() not in SW_DatesandNumbers])
    txt3=' '.join([word for word in txt2.split() if word.lower() not in SW_Generic])
    txt4=' '.join([word for word in txt3.split() if word.lower() not in SW_GenericLong])
    txt5=' '.join([word for word in txt4.split() if word.lower() not in SW_Geographic])
    return ' '.join([word for word in txt5.split() if word.lower() not in SW_Names])


In [ ]:
#applying func for each row
df['data']=df['data'].apply(text_process)

In [ ]:
#importing master Dictionary
positive=pd.read_csv("/content/drive/MyDrive/20211030 Test Assignment/MasterDictionary/positive-words.txt",header=None)
negative=pd.read_csv("/content/drive/MyDrive/20211030 Test Assignment/MasterDictionary/negative-words.txt",header=None,encoding="ISO-8859-1",sep='\n' )


In [ ]:
positive.columns=['pos']
negative.columns=['neg']
positive['pos']=positive['pos'].astype(str)
negative['neg']=negative['neg'].astype(str)

In [ ]:
#positive and negative dictionary without stopwords
positive['pos']=positive['pos'].apply(text_process)
negative['neg']=negative['neg'].apply(text_process)

In [ ]:
#positive list
length=positive.shape[0]
pos=[]
for i in range(0,length):
   nopunc =[char for char in positive.iloc[i] if char not in string.punctuation or char != '+']
   nopunc=''.join(nopunc)

   pos.append(nopunc)

In [ ]:
#negative list
length=negative.shape[0]
neg=[]
for i in range(0,length):
  nopunc =[char for char in negative.iloc[i] if char not in string.punctuation or char != '+']
  nopunc=''.join(nopunc)
  neg.append(nopunc)

# TOKENIZATION

In [ ]:
#importing tokenize library
from nltk.tokenize import word_tokenize

In [ ]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
df['tokenized_sents'] = df.apply(lambda row: nltk.word_tokenize(row['data']), axis=1)

In [ ]:
df['tokenized_sents'].head()

0    [Ranking, customer, behaviours, for, business,...
1    [Ranking, customer, behaviours, for, business,...
2    [Ranking, customer, behaviours, for, business,...
3    [Ranking, customer, behaviours, for, business,...
4    [Ranking, customer, behaviours, for, business,...
Name: tokenized_sents, dtype: object

In [ ]:
df['total_len'] = df['tokenized_sents'].map(lambda x: len(x))

# CALCULATING SCORES

### 1) POSITIVE SCORE

In [ ]:
num_pos = df['tokenized_sents'].map(lambda x: len([i for i in x if i in pos]))
df['positive_score'] = num_pos


### 2) NEGATIVE SCORE

In [ ]:
num_neg = df['tokenized_sents'].map(lambda x: len([i for i in x if i in neg]))
df['negative_score'] = num_neg

### 3) POLARITY SCORE

In [ ]:
#Polarity Score formula = (Positive Score – Negative Score)/ ((Positive Score + Negative Score) + 0.000001)
Polarity_Score=(num_pos-num_neg)/((num_pos+num_neg)+0.000001)
df['Polarity_Score'] = Polarity_Score

### 4) SUBJECTIVITY SCORE

In [ ]:
#Subjectivity Score formula = (Positive Score + Negative Score)/ ((Total Words after cleaning) + 0.000001)
subjectiivity_score=(num_pos-num_neg)/((len(df['tokenized_sents']))+ 0.000001)
df['subjectiivity_score'] = subjectiivity_score

### 5) AVG SENTENCE LENGTH

In [ ]:
sum_of_length_of_sents = sum(df['total_len'].values)

In [ ]:
sum_of_length_of_sents

184976

In [ ]:
length_of_sentence = []
for i in df['total_len']:
  length_of_sentence.append(i)

In [ ]:
avg_sents_length=sum_of_length_of_sents / length_of_sentence
df['avg_senetence_length'] = avg_sents_length

### 6) PERCENTAGE OF COMPLEX WORDS

In [ ]:
#complex word individual count
vowels=['a','e','i','o','u']
import re
count=0
complex_Word_Count=0
for i in str(df['tokenized_sents']):
  x=re.compile('[es|ed]$')
  if x.match(i.lower()):
   count+=0
  else:
    for j in str(df['tokenized_sents']):
      if(j.lower() in vowels ):
        count+=1
  if(count>2):
   complex_Word_Count+=1
  count=0


In [ ]:
Percentage_of_Complex_words=complex_Word_Count/df['total_len']
df['Percentage_of_Complex_words'] = Percentage_of_Complex_words

### 7) FOG INDEX

In [ ]:
#Fog Index formula = 0.4 * (Average Sentence Length + Percentage of Complex words)
Fog_Index = 0.4 * (avg_sents_length + Percentage_of_Complex_words)
df['Fog_Index'] = Fog_Index

### 8) AVG NUMBER OF WORDS PER SENTENCE

In [ ]:

avg_no_of_words_per_sentence=df['total_len']/ df.shape[0]
df['avg_no_of_words_per_sentence'] = avg_no_of_words_per_sentence

### 9) COMPLEX WORD COUNT

In [ ]:
def count_complex_words(word):  
  vowels=['a','e','i','o','u']
  import re
  count=0
  complex_Word_Count=0
  for i in word:
    x=re.compile('[es|ed]$')
    if x.match(i.lower()):
     count+=0
    else:
      for j in i:
        if(j.lower() in vowels ):
          count+=1
    if(count>2):
     complex_Word_Count+=1
    count=0
  return complex_Word_Count

In [ ]:
df['complex_Word_Count'] = df['tokenized_sents'].apply(lambda x: count_complex_words((x)))

In [ ]:
df.head()

URL  \
0  https://insights.blackcoffer.com/challenges-an...   
1  https://insights.blackcoffer.com/business-anal...   
2  https://insights.blackcoffer.com/big-data-anal...   
3  https://insights.blackcoffer.com/the-future-of...   
4  https://insights.blackcoffer.com/blockchain-fo...   

                                                data  \
0  Ranking customer behaviours for business strat...   
1  Ranking customer behaviours for business strat...   
2  Ranking customer behaviours for business strat...   
3  Ranking customer behaviours for business strat...   
4  Ranking customer behaviours for business strat...   

                                     tokenized_sents  total_len  \
0  [Ranking, customer, behaviours, for, business,...       1118   
1  [Ranking, customer, behaviours, for, business,...        766   
2  [Ranking, customer, behaviours, for, business,...       1254   
3  [Ranking, customer, behaviours, for, business,...       1667   
4  [Ranking, customer, behaviours, for, business,...        983   

   positive_score  negative_score  Polarity_Score  subjectiivity_score  \
0              45              39        0.071429             0.041096   
1              33               4        0.783784             0.198630   
2              30              44       -0.189189            -0.095890   
3              47              12        0.593220             0.239726   
4              29              28        0.017544             0.006849   

   avg_senetence_length  Percentage_of_Complex_words  Fog_Index  \
0            165.452594                     0.523256  66.390340   
1            241.483029                     0.763708  96.898695   
2            147.508772                     0.466507  59.190112   
3            110.963407                     0.350930  44.525735   
4            188.174975                     0.595117  75.508037   

   avg_no_of_words_per_sentence  complex_Word_Count  
0                      7.657534                 279  
1                      5.246575                 259  
2                      8.589041                 366  
3                     11.417808                 454  
4                      6.732877                 265

### 10) WORD COUNT

In [ ]:
df['word_count'] = df['tokenized_sents'].map(lambda x: len(x))

### 11) SYLLABLE PER WORD

In [ ]:
def syllable_count(word):
    word = word.lower()
    count = 0
    vowels = "aeiouy"
    if word[0] in vowels:
        count += 1
    for index in range(1, len(word)):
        if word[index] in vowels and word[index - 1] not in vowels:
            count += 1
    if word.endswith("e"):
        count -= 1
    if count == 0:
        count += 1
    return count

In [ ]:
df['syllable_count'] = df['tokenized_sents'].apply(lambda x: syllable_count(str(x)))

### 12) PERSONAL PRONOUNS

In [ ]:
def pronoun_count(word):  
  pronouns=['i','we','my','ours','us' ]
  import re
  count=0
  for i in word:
    if i.lower() in pronouns:
      count+=1
  return count

In [ ]:
df['personal_pronouns'] = df['tokenized_sents'].apply(lambda x: pronoun_count((x)))

In [ ]:
df.head()

URL  \
0  https://insights.blackcoffer.com/challenges-an...   
1  https://insights.blackcoffer.com/business-anal...   
2  https://insights.blackcoffer.com/big-data-anal...   
3  https://insights.blackcoffer.com/the-future-of...   
4  https://insights.blackcoffer.com/blockchain-fo...   

                                                data  \
0  Ranking customer behaviours for business strat...   
1  Ranking customer behaviours for business strat...   
2  Ranking customer behaviours for business strat...   
3  Ranking customer behaviours for business strat...   
4  Ranking customer behaviours for business strat...   

                                     tokenized_sents  total_len  \
0  [Ranking, customer, behaviours, for, business,...       1118   
1  [Ranking, customer, behaviours, for, business,...        766   
2  [Ranking, customer, behaviours, for, business,...       1254   
3  [Ranking, customer, behaviours, for, business,...       1667   
4  [Ranking, customer, behaviours, for, business,...        983   

   positive_score  negative_score  Polarity_Score  subjectiivity_score  \
0              45              39        0.071429             0.041096   
1              33               4        0.783784             0.198630   
2              30              44       -0.189189            -0.095890   
3              47              12        0.593220             0.239726   
4              29              28        0.017544             0.006849   

   avg_senetence_length  Percentage_of_Complex_words  Fog_Index  \
0            165.452594                     0.523256  66.390340   
1            241.483029                     0.763708  96.898695   
2            147.508772                     0.466507  59.190112   
3            110.963407                     0.350930  44.525735   
4            188.174975                     0.595117  75.508037   

   avg_no_of_words_per_sentence  complex_Word_Count  word_count  \
0                      7.657534                 279        1118   
1                      5.246575                 259         766   
2                      8.589041                 366        1254   
3                     11.417808                 454        1667   
4                      6.732877                 265         983   

   syllable_count  personal_pronouns  
0            1957                  8  
1            1476                  0  
2            2162                  2  
3            2793                 18  
4            1712                  9

### 13) AVG WORD LENGTH

In [ ]:
def word_count_number(sentence):
  count=0
  for i in (sentence):
    for j in i:
      for k in j:
        count += 1
  return count

In [ ]:
df['number_of_words'] = df['tokenized_sents'].apply(lambda x: word_count_number((x)))

In [ ]:
df['avg_word_length'] = df['number_of_words']/df['word_count']

In [ ]:
df.drop(columns = (['number_of_words']), inplace = True)

In [ ]:
output_data=df.copy()

In [ ]:
output_data.head()

URL  \
0  https://insights.blackcoffer.com/challenges-an...   
1  https://insights.blackcoffer.com/business-anal...   
2  https://insights.blackcoffer.com/big-data-anal...   
3  https://insights.blackcoffer.com/the-future-of...   
4  https://insights.blackcoffer.com/blockchain-fo...   

                                                data  \
0  Ranking customer behaviours for business strat...   
1  Ranking customer behaviours for business strat...   
2  Ranking customer behaviours for business strat...   
3  Ranking customer behaviours for business strat...   
4  Ranking customer behaviours for business strat...   

                                     tokenized_sents  total_len  \
0  [Ranking, customer, behaviours, for, business,...       1118   
1  [Ranking, customer, behaviours, for, business,...        766   
2  [Ranking, customer, behaviours, for, business,...       1254   
3  [Ranking, customer, behaviours, for, business,...       1667   
4  [Ranking, customer, behaviours, for, business,...        983   

   positive_score  negative_score  Polarity_Score  subjectiivity_score  \
0              45              39        0.071429             0.041096   
1              33               4        0.783784             0.198630   
2              30              44       -0.189189            -0.095890   
3              47              12        0.593220             0.239726   
4              29              28        0.017544             0.006849   

   avg_senetence_length  Percentage_of_Complex_words  Fog_Index  \
0            165.452594                     0.523256  66.390340   
1            241.483029                     0.763708  96.898695   
2            147.508772                     0.466507  59.190112   
3            110.963407                     0.350930  44.525735   
4            188.174975                     0.595117  75.508037   

   avg_no_of_words_per_sentence  complex_Word_Count  word_count  \
0                      7.657534                 279        1118   
1                      5.246575                 259         766   
2                      8.589041                 366        1254   
3                     11.417808                 454        1667   
4                      6.732877                 265         983   

   syllable_count  personal_pronouns  avg_word_length  
0            1957                  8         4.812165  
1            1476                  0         5.449086  
2            2162                  2         4.811005  
3            2793                 18         4.901020  
4            1712                  9         5.109868

In [ ]:
output_data.drop(columns = (['data','tokenized_sents', 'total_len' ]), inplace = True)

In [ ]:
output_data.head()

URL  positive_score  \
0  https://insights.blackcoffer.com/challenges-an...              45   
1  https://insights.blackcoffer.com/business-anal...              33   
2  https://insights.blackcoffer.com/big-data-anal...              30   
3  https://insights.blackcoffer.com/the-future-of...              47   
4  https://insights.blackcoffer.com/blockchain-fo...              29   

   negative_score  Polarity_Score  subjectiivity_score  avg_senetence_length  \
0              39        0.071429             0.041096            165.452594   
1               4        0.783784             0.198630            241.483029   
2              44       -0.189189            -0.095890            147.508772   
3              12        0.593220             0.239726            110.963407   
4              28        0.017544             0.006849            188.174975   

   Percentage_of_Complex_words  Fog_Index  avg_no_of_words_per_sentence  \
0                     0.523256  66.390340                      7.657534   
1                     0.763708  96.898695                      5.246575   
2                     0.466507  59.190112                      8.589041   
3                     0.350930  44.525735                     11.417808   
4                     0.595117  75.508037                      6.732877   

   complex_Word_Count  word_count  syllable_count  personal_pronouns  \
0                 279        1118            1957                  8   
1                 259         766            1476                  0   
2                 366        1254            2162                  2   
3                 454        1667            2793                 18   
4                 265         983            1712                  9   

   avg_word_length  
0         4.812165  
1         5.449086  
2         4.811005  
3         4.901020  
4         5.109868

#This is the final dataset which contains all scores for the respective URLs